File: src/communication/Unmarshalling.java
package SC6103_DS.src.communication;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

public class Unmarshalling {

    // Demarshal an entire Message object from byte array
    public static Message demarshalMessage(byte[] byteArray) {
        ByteBuffer buffer = ByteBuffer.wrap(byteArray);

        byte messageType = buffer.get();
        int requestId = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        int dataLength = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });

        byte[] data = new byte[dataLength];
        buffer.get(data);

        return new Message(messageType, requestId, dataLength, data);
    }

    // Demarshal a Flight object from byte array
    public static Message.Flight demarshalFlight(byte[] byteArray) {
        ByteBuffer buffer = ByteBuffer.wrap(byteArray);

        int flightId = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        String sourcePlace = demarshalString(buffer);
        String destinationPlace = demarshalString(buffer);
        Message.DepartureTime departureTime = demarshalDepartureTime(buffer);
        float airfare = demarshalFloat(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        int seatAvailability = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        int baggageAvailability = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });

        return new Message.Flight(flightId, sourcePlace, destinationPlace, departureTime, airfare, seatAvailability, baggageAvailability);
    }

    // Demarshal a DepartureTime object from ByteBuffer
    public static Message.DepartureTime demarshalDepartureTime(ByteBuffer buffer) {
        int year = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        int month = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        int day = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        int hour = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        int minute = demarshalInt(new byte[] { buffer.get(), buffer.get(), buffer.get(), buffer.get() });
        return new Message.DepartureTime(year, month, day, hour, minute);
    }

    // Method to demarshal a String from ByteBuffer
    public static String demarshalString(ByteBuffer buffer) {
        int length = buffer.getInt();
        byte[] stringBytes = new byte[length];
        buffer.get(stringBytes);
        return new String(stringBytes, StandardCharsets.UTF_8);
    }

    // Method to demarshal an int from byte array
    public static int demarshalInt(byte[] data) {
        return ByteBuffer.wrap(data).getInt();
    }

    // Method to demarshal a float from byte array
    public static float demarshalFloat(byte[] data) {
        return ByteBuffer.wrap(data).getFloat();
    }
}

File: src/communication/unmarshalling.c
#include "communication.h"
#include <stdlib.h>
#include <string.h>

// Unmarshal an integer from a byte array
int unmarshal_int(const uint8_t* byte_array, uint32_t* offset) {
    uint32_t network_value;
    memcpy(&network_value, byte_array + *offset, 4);
    *offset += 4;
    return ntohl(network_value);
}

// Unmarshal a float from a byte array
float unmarshal_float(const uint8_t* byte_array, uint32_t* offset) {
    uint32_t network_value;
    memcpy(&network_value, byte_array + *offset, 4);
    *offset += 4;
    network_value = ntohl(network_value);
    return *(float*)&network_value;
}

// Unmarshal a string from a byte array
char* unmarshal_string(const uint8_t* byte_array, uint32_t* offset) {
    uint32_t str_len = unmarshal_int(byte_array, offset);
    char* str = malloc(str_len + 1);
    memcpy(str, byte_array + *offset, str_len);
    str[str_len] = '\0';
    *offset += str_len;
    return str;
}

// Unmarshal a DepartureTime structure from a byte array
DepartureTime* unmarshal_departure_time(const uint8_t* byte_array, uint32_t* offset) {
    DepartureTime* time = malloc(sizeof(DepartureTime));
    time->year = unmarshal_int(byte_array, offset);
    time->month = unmarshal_int(byte_array, offset);
    time->day = unmarshal_int(byte_array, offset);
    time->hour = unmarshal_int(byte_array, offset);
    time->minute = unmarshal_int(byte_array, offset);
    return time;
}

// Unmarshal a Flight structure from a byte array
Flight* unmarshal_flight(const uint8_t* byte_array, uint32_t* offset) {
    Flight* flight = malloc(sizeof(Flight));
    flight->flight_id = unmarshal_int(byte_array, offset);
    flight->source_place = unmarshal_string(byte_array, offset);
    flight->destination_place = unmarshal_string(byte_array, offset);
    flight->departure_time = unmarshal_departure_time(byte_array, offset);
    flight->airfare = unmarshal_float(byte_array, offset);
    flight->seat_availability = unmarshal_int(byte_array, offset);
    flight->baggage_availability = unmarshal_int(byte_array, offset);
}